tree/dataframe/inc/ROOT/RDF/RMergeableValue.hxx

```diff
@@ -0,0 +1,128 @@
+
+namespace ROOT {
+namespace RDF {
+
+template <typename T>
+class RMergeableValue {
+protected:
+   T fValue;
+
+public:
+   RMergeableValue(const T &value) : fValue{value} {}
+   virtual ~RMergeableValue() = default;
+   virtual void Merge(const std::unique_ptr<RMergeableValue<T>> &) = 0;
+   virtual const T &GetValue() const { return fValue; }
+};
+
+// Specialization for generic action
+template <typename T>
+class RMergeableGenericAction final : public RMergeableValue<T> {
+   using mergefun_t = std::function<T(const T &, const T &)>;
+   const mergefun_t fMergeFun;
+
+public:
+   RMergeableGenericAction(const T &value, const mergefun_t &mergefun) : RMergeableValue<T>(value), fMergeFun{mergefun}
+   {
+   }
+
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final
+   {
+      const auto &othervalue = other->GetValue();
+      this->fValue = fMergeFun(this->fValue, othervalue);
+   }
+};
+
+// RMergeableValue specialization for the Mean action
+template <typename T>
+class RMergeableMean final : public RMergeableValue<T> {
+   ULong64_t fCounts;
+
+public:
+   RMergeableMean(const T &value, const ULong64_t &counts) : RMergeableValue<T>(value), fCounts{counts} {}
+
+   const ULong64_t &GetCounts() const { return fCounts; }
+
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final
+   {
+      const auto &othervalue = other->GetValue();
+      const auto &othercounts = dynamic_cast<const RMergeableMean<T> &>(*other).GetCounts();
+      const auto num = this->fValue * fCounts + othervalue * othercounts;
+      const auto denum = static_cast<double>(fCounts + othercounts);
+      this->fValue = num / denum;
+      fCounts += othercounts;
+   }
+};
+
+template <typename T>
+class RMergeableHisto1D final : public RMergeableValue<T> {
+public:
+   RMergeableHisto1D(const T &value) : RMergeableValue<T>(value) {}
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final { this->fValue.Add(&(other->GetValue())); }
+};
+
+template <typename T>
+class RMergeableFillParHelper final : public RMergeableValue<T> {
+public:
+   RMergeableFillParHelper(const T &value) : RMergeableValue<T>(value) {}
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final
+   {
+      TList l;
+      l.Add(const_cast<T *>(&(other->GetValue())));
+      this->fValue.Merge(&l);
+   }
+};
+
+// RMergeableValue specialization for the StdDev action
+template <typename T>
+class RMergeableStdDev final : public RMergeableValue<T> {
+   ULong64_t fCounts;
+   Double_t fMean;
+
+public:
+   RMergeableStdDev(const T &value, const ULong64_t &counts, const Double_t &mean)
+      : RMergeableValue<T>(value), fCounts{counts}, fMean{mean}
+   {
+   }
+
+   const ULong64_t &GetCounts() const { return fCounts; }
+
+   const Double_t &GetMean() const { return fMean; }
+
+   void Merge(const std::unique_ptr<RMergeableValue<T>> &other) final
+   {
+      const auto &othermergeable = dynamic_cast<const RMergeableStdDev<T> &>(*other);
```

Comment:
can be a static_cast if you don't check for null

Reply:
Kept it `dynamic_cast` and checked for `std::bad_cast` errors

Reply:
humm .. where is the check for std::bad_cast?  Why is catching an exception better than doing a cast to pointer and checking for nullptr?

Reply:
I'm sorry, this conversation is a bit outdated. The check for std::bad_cast is present in the latest commit I pushed (with a bit of git rebase though).

As for the cast to pointer, I really have no strong opinion. I just preferred keeping it as reference to match the value category of the function parameter. But if you think that brings a strong disadvantage I can change it